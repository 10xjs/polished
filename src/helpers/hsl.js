// @flow

import reduceHexValue from '../internalHelpers/_reduceHexValue'
import toHex from '../internalHelpers/_numberToHex'

type HslColor = {
  hue: number,
  saturation: number,
  lightness: number,
}

/**
 * Returns a string value for the color. The returned result is the smalles possible hex notation.
 *
 * @example
 * // Styles as object usage
 * const styles = {
 *   background: hsl(360, 0.75, 0.4),
 *   background: hsl({ hue: 360, saturation: 0.75, lightness: 0.4 }),
 * }
 *
 * // styled-components usage
 * const div = styled.div`
 *   background: ${hsl(360, 0.75, 0.4)};
 *   background: ${hsl({ hue: 360, saturation: 0.75, lightness: 0.4 })};
 * `
 *
 * // CSS in JS Output
 *
 * element {
 *   background: "#b3191c";
 *   background: "#b3191c";
 * }
 */

function hslToHex(hue: number, saturation: number, lightness: number): string {
  if (saturation === 0) {
    // achromatic
    return reduceHexValue(`#${toHex(lightness)}${toHex(lightness)}${toHex(lightness)}`)
  }

  // formular from https://en.wikipedia.org/wiki/HSL_and_HSV
  const huePrime = hue / 60
  const chroma = (1 - Math.abs((2 * lightness) - 1)) * saturation
  const secondComponent = chroma * (1 - Math.abs((huePrime % 2) - 1))

  let red
  let green
  let blue
  if (huePrime >= 0 && huePrime <= 1) {
    red = chroma
    green = secondComponent
    blue = 0
  } else if (huePrime >= 1 && huePrime <= 2) {
    red = secondComponent
    green = chroma
    blue = 0
  } else if (huePrime >= 2 && huePrime <= 3) {
    red = 0
    green = chroma
    blue = secondComponent
  } else if (huePrime >= 3 && huePrime <= 4) {
    red = 0
    green = secondComponent
    blue = chroma
  } else if (huePrime >= 4 && huePrime <= 5) {
    red = secondComponent
    green = 0
    blue = chroma
  } else if (huePrime >= 5 && huePrime < 6) {
    red = chroma
    green = 0
    blue = secondComponent
  }

  const lightnessModification = lightness - (chroma / 2)
  const addLightAndConvert = (color) => toHex(Math.round((color + lightnessModification) * 255))
  return reduceHexValue(`#${addLightAndConvert(red)}${addLightAndConvert(green)}${addLightAndConvert(blue)}`)
}

function hsl(value: HslColor | number, saturation?: number, lightness?: number): string {
  if (typeof value === 'number' && typeof saturation === 'number' && typeof lightness === 'number') {
    return hslToHex(value, saturation, lightness)
  } else if (typeof value === 'object' && saturation === undefined && lightness === undefined) {
    return hslToHex(value.hue, value.saturation, value.lightness)
  }

  throw new Error('Passed invalid arguments to hsl, please pass multiple numbers e.g. hsl(360, 0.75, 0.4) or an object e.g. rgb({ hue: 255, saturation: 0.4, lightness: 0.75 }).')
}

export default hsl
