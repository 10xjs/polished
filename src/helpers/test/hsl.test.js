// @flow
import hsl from '../hsl'

describe('hsl', () => {
  it('should convert numbers to a hex color', () => {
    expect({ background: hsl(360, 0.75, 0.4) }).toMatchSnapshot()
  })

  it('should convert the color red', () => {
    expect({ background: hsl(0, 1, 0.5) }).toMatchSnapshot()
  })

  it('should convert the color yellow', () => {
    expect({ background: hsl(60, 1, 0.5) }).toMatchSnapshot()
  })

  it('should convert the color lime', () => {
    expect({ background: hsl(120, 1, 0.5) }).toMatchSnapshot()
  })

  it('should convert the color cyan', () => {
    expect({ background: hsl(180, 1, 0.5) }).toMatchSnapshot()
  })

  it('should convert the color blue', () => {
    expect({ background: hsl(240, 1, 0.5) }).toMatchSnapshot()
  })

  it('should convert the color magenta', () => {
    expect({ background: hsl(300, 1, 0.5) }).toMatchSnapshot()
  })

  it('should convert black', () => {
    expect({ background: hsl({ hue: 360, saturation: 0, lightness: 0.4 }) }).toMatchSnapshot()
  })

  it('should convert a hls object to a hex color', () => {
    expect({ background: hsl({ hue: 360, saturation: 0.75, lightness: 0.4 }) }).toMatchSnapshot()
  })

  it('should convert to a reduce hex value if possible', () => {
    expect({ background: hsl(360, 1, 0.4) }).toMatchSnapshot()
  })

  it('should throw an error if an object and multiple arguments are passed', () => {
    expect(() => ({ background: hsl({ hue: 360, saturation: 0.75, lightness: 0.4 }, 250, 100) }))
      .toThrow('Passed invalid arguments to hsl, please pass multiple numbers e.g. hsl(360, 0.75, 0.4) or an object e.g. rgb({ hue: 255, saturation: 0.4, lightness: 0.75 }).')
  })
})
