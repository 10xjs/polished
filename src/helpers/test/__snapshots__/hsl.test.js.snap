exports[`hsl should convert a hls object to a hex color 1`] = `
Object {
  "background": "#b31919",
}
`;

exports[`hsl should convert black 1`] = `
Object {
  "background": "#666",
}
`;

exports[`hsl should convert numbers to a hex color 1`] = `
Object {
  "background": "#b31919",
}
`;

exports[`hsl should convert the color blue 1`] = `
Object {
  "background": "#00f",
}
`;

exports[`hsl should convert the color cyan 1`] = `
Object {
  "background": "#0ff",
}
`;

exports[`hsl should convert the color lime 1`] = `
Object {
  "background": "#0f0",
}
`;

exports[`hsl should convert the color magenta 1`] = `
Object {
  "background": "#f0f",
}
`;

exports[`hsl should convert the color red 1`] = `
Object {
  "background": "#f00",
}
`;

exports[`hsl should convert the color yellow 1`] = `
Object {
  "background": "#ff0",
}
`;

exports[`hsl should convert to a reduce hex value if possible 1`] = `
Object {
  "background": "#c00",
}
`;
